File: app/src/main/java/com/example/mobile6/ui/base/BaseFragment.java
package com.example.mobile6.ui.base;

import static androidx.navigation.fragment.FragmentKt.findNavController;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewbinding.ViewBinding;

import java.util.function.Consumer;

public abstract class BaseFragment<VB extends ViewBinding> extends Fragment {
    protected VB binding;
    protected Bundle args;

    protected abstract VB inflateBinding(LayoutInflater layoutInflater, ViewGroup container);

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        binding = inflateBinding(inflater, container);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        args = getArguments();
        initViews();
    }

    protected void initViews() {

    }

    protected void back() {
        findNavController(this).navigateUp();
    }

    protected void back(String requestKey, Bundle bundle) {
        getParentFragmentManager().setFragmentResult(requestKey, bundle);
        back();
    }

    protected void setOnBackResultListener(String requestKey, Consumer<Bundle> consumer) {
        getParentFragmentManager().setFragmentResultListener(requestKey, this, (requestKey1, result) -> {
            consumer.accept(result);
        });
    }

    protected void navigateTo(int action, @Nullable Bundle bundle) {
        findNavController(this).navigate(action, bundle);
    }

    protected void navigateTo(int action) {
        findNavController(this).navigate(action);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/mobile6/presentation/ui/base/BaseFragment.kt
package com.example.mobile6.presentation.ui.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavOptions
import androidx.navigation.fragment.findNavController
import androidx.viewbinding.ViewBinding
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

abstract class BaseFragment<VB : ViewBinding> : Fragment() {
    private var _binding: VB? = null
    protected val binding get() = _binding!!

    protected abstract val bindingInflater: (layoutInflater: LayoutInflater, container: ViewGroup?) -> VB

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = bindingInflater.invoke(inflater, container)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        arguments?.let { processArguments(it) }
        initViews()
        initObservers()
    }

    abstract fun initViews()

    abstract fun initObservers()

    protected open fun processArguments(args: Bundle) {

    }

    protected fun back() {
        findNavController().navigateUp()
    }

    protected fun <T> back(requestKey: String, value: T) {
        findNavController().previousBackStackEntry?.savedStateHandle?.set(requestKey, value)
        back()
    }

    protected fun <T> observeBackResult(requestKey: String, onResult: (T) -> Unit) {
        findNavController().currentBackStackEntry?.savedStateHandle?.getStateFlow<T?>(
            requestKey,
            null
        )
            ?.onEach {
                it?.let { value ->
                    onResult(value)
                }
            }?.flowWithLifecycle(viewLifecycleOwner.lifecycle, Lifecycle.State.STARTED)
            ?.launchIn(viewLifecycleOwner.lifecycleScope)
    }

    protected fun navigateTo(
        actionId: Int,
        args: Bundle? = null,
        navOptions: NavOptions? = null
    ) {
        findNavController().navigate(actionId, args, navOptions)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}